### DISCUSSION TITLE
QQuickImageProvider equivalent in slint?

### DISCUSSION THREAD

**Discussion by bog-dan-ro** [slint-ui/slint#2712] [General] [answered]
Hi,

What's the best way to provide images to slint Image ui element? 
Is there any QQuickImageProvider equivalent in slint?
Slint should cache the image texture by some id/path and reuse it when needed.

**Comment by tronical** [comment, marked as answer]
Images are specified with the `source` property. You can either supply that give the literal `@image-url("string literal path to image here")` or you can programmatically create `slint::Image` objects in [Rust](https://slint-ui.com/releases/1.0.2/docs/cpp/api/structslint_1_1image) and [C++](https://slint-ui.com/releases/1.0.2/docs/cpp/api/structslint_1_1image) and assign them to the `source` property. The type of the `source` property is `image`, which maps to `slint::Image`.

The documentation linked above has examples how to create `slint::Image` objects from pixel buffers. Those are shared by reference count. Both, in Rust and C++, `slint::Image` can also be created from a path to an image file on disk, using `load_from_path()`.

If you repeatedly call `slint::Image::load_from_path()` with the same path, it will only load the image from disk the first time. Subsequent calls will re-use the previously loaded image from a cache. The cache has a capacity of roughly 5MB at the time of writing, the size of each item in the cache is weighed by the weight of the image in bytes (where possible).

Once a `slint::Image` is assigned to the `source` property of an `Image` element, the renderer takes care of allocating and populating a texture for it (unless it's the software renderer). That texture is cached by the same keying mechanism as the regular image cache, so the texture will not be re-populated on every frame.

**Reply by bog-dan-ro** [reply]
Thanks!

**Reply by bog-dan-ro** [reply]
If I'm creating 100 `slint::Image` from the same `SharedPixelBuffer` will they share the same texture?

**Reply by tronical** [reply]
No, such images will not share the same texture. It might be possible to fix this, I'm not sure.

**Reply by bog-dan-ro** [reply]
Well, without sharing the same texture it will waste a lot of resources ;-)
Maybe turn this discussion into a task?

**Reply by tronical** [reply]
Sure.
